dill/_dill.py

```diff
@@ -437,8 +373,7 @@ def _stash_modules(main_module):
     imported_as = []
     imported_top_level = []  # keep separeted for backwards compatibility
     original = {}
-    items = 'items' if PY3 else 'iteritems'
-    for name, obj in getattr(main_module.__dict__, items)():
+    for name, obj in vars(main_module).items():
```

Comment:
Not sure if this should stay `main_module.__dict__` as opposed to `vars(main_module)`.  They are not always the same.

Reply:
According to [documentation](https://docs.python.org/3/library/functions.html#vars) (including from Py2.7), it _is_ always the case:

> `vars([object])`
> Return the `__dict__` attribute for a module, class, instance, or any other object with a `__dict__` attribute.
> (...)
> A `TypeError` exception is raised if an object is specified but it doesn’t have a `__dict__` attribute (...).

This is good if we have to distinguish a missing `__dict__` from an `AttributeError` in the same code block. However, just above the `vars()` documentation is `type()` documentation, and it states:

> `class type(object)`
> (...)
> With one argument, return the type of an object. The return value is a type object and **generally** the same object as returned by `object.__class__`.

Although I don't know what _that_ means...

Reply:
By the way, I think I changed only this instance to match what I used in other session functions.
 
Regardless of the type of exception they raise on failure, `obj.__dict__` and `vars(obj)` can be considered just different styles. Both are used extensively in the code base: `_dill` and `_objects` tend to use the former, while the modules `detect` and `source` (and the new `session` module in #475 ) use mostly the latter. It would be good to standardize this.

I like the fact that `vars(module)` is close in syntax to `globals()`, and also the distinct error type 
—that I wasn't aware of actually.

Reply:
If you look at the part you converted to (...), I believe that `vars` acts more like `locals()` than `globals()` (so it can be masked).

Reply:
Of course, `vars()` without arguments is equivalent to `locals()`, and the restrictions that apply to `locals()` also apply **in this case.** But for `vars(obj)` the behavior is strictly equivalent to a function defined as:

```python
def vars(obj):
    if not hasattr(obj, '__dict__'):
        raise TypeError
    return obj.__dict__
```

The implementation in CPython is just like it: https://github.com/python/cpython/blob/bbcf42449e13c0b62f145cd49d12674ef3d5bf64/Python/bltinmodule.c#L2404-L2423

Reply:
So I'll update everything to `vars(module)`, OK?

Reply:
I wouldn't unless you are getting a speed bonus out of it -- otherwise the change just seems to be semantic sugar.  (I'm also a bit hesitant about some of the error specification you did as you might be able to tell from my comments.)

Reply:
> I wouldn't unless you are getting a speed bonus out of it -- otherwise the change just seems to be semantic sugar.

As you wish. And oops! `vars(module)` is actually slower... OK, so I won't change anything, but I'll try to keep consistency with the submodule style in future PRs.

> (I'm also a bit hesitant about some of the error specification you did as you might be able to tell from my comments.)
Yes, I'm reviewing these carefully. I really wasn't sure about some of the specific errors, but I left them on purpose for you to review and eventually clarify if they were appropriate. If we can't agree on whether a exception list cover all the expected cases (not including misuses that should rise an error), I'll change it to `except Exception`, as I did with other `try` blocks for which there was no clue of the possible error types.

> There's a lot of changes in this PR, so we need to make sure this gets well tested.
Yeah... I may have diverted a bit from the main goals. Let's wrap it up.

Reply:
> I'll change it to except Exception, as I did with other try blocks for which there was no clue of the possible error types.

This is better... as it's a change in how something pickles if the error is no longer caught, but maybe gets caught elsewhere.